### DISCUSSION TITLE
Some Clarifications Regarding aIn records

### DISCUSSION THREAD

**Discussion by Robatronic** [dls-controls/pythonSoftIOC#72] [Q&A] [answered]
It seems there's a variety of ways to define ioc records and I'm looking to confirm or clarify my understanding thus far.

Two main categories are ones using the special PythonIOC device and the lower level ones in builder.records.

I defined:
```python
ai2 = builder.records.ai('AI2', VAL="5", INP="float1 CP MS", SCAN="Passive", PINI="YES")
```
and updates were seen with a camonitor running in another console.

Conversely:
```python
ai1 = builder.aIn('AI', initial_value=5, INP="float1 CP MS", SCAN="Passive", PINI="YES")
```
did not update (verified using both caget and camonitor from another console)

So, I set up a monitor on float1 (as per the tutorials) and wrote the value from the callback function. Updates still did not go out. I then changed the SCAN to "I/O Intr" for ai2 and updates started going out.

Do all these "higher-level" records require SCAN="I/O Intr"?
Do they also require that the value is  manually set in Python (i.e. ai.set(value))......?

Or should I be updated ai1 via an ao1's on_update handler....? Is there a "best practice" for what I'm attempting?

Thanks so much for this intriguing tool/utility!!! It holds a lot of promise for us here.  :)

**Comment by Robatronic** [comment]
Is PINI meaningful for PVs in this category...?

Sorry for all the questions. Trying to get my head around some of the (what seems to be) nuances.

**Comment by AlexanderWells-diamond** [comment]
I should probably leave a full explanation to @Araneidae or @thomascobb , but I can give direct answers based on current code to some of your questions. I'll leave it for others to say if this behaviour is correct and/or could be improved.

By default all `builder.*In` records (i.e. every In record type done through the top level API) are created with `SCAN='I/O Intr'`. (See `_in_record(...)` in  `builder.py`).
Immediately following this, if an `initial_value` is specified `PINI` will be set to `YES`.

It sounds like you may have encountered issue #67 , or some variant thereof, if you are seeing records not update in some cases.

If it's helpful, you can print out the record database file using `builder.WriteRecords("my_records.db")`. Examining the resultant file may help you understand what pythonSoftIOC is doing automatically for you.

**Comment by Araneidae** [comment, marked as answer]
The other piece of the puzzle to understand is the `DTYP` field and how EPICS record processing works.
Every EPICS record has "device support" associated with it: this determines how the fields in the record behave when the record is processed.  Device support is implemented in C (or using `ctypes` bindings in pythonSoftIOC) and is bound to each individual EPICS record through the `DTYP` field.

For records created with the core `builder.records` constructors the `DTYP` field is left blank, which means that record support is entirely passive.  In other words, when a record processes it does nothing.  On the other hand, all the `builder.*In` and `builder.*Out` constructors set `DTYP` to the bind the record processing to pythonSoftIOC processing.  At the same time, a number of other fields are set.

When a pythonSoftIOC IN record processes the record support code (implemented in `device.py`) copies the value set in the record (by specifying `initial_value` or by calling the `.set()` method) from Python into the record's `.VAL` field.  Thus when an IN record processes its value will always be the value set in software, and doing caput to an IN record is futile!  The `SCAN` field for IN records is set to `I/O Intr` so that calls to `.set()` can trigger record processing.

For OUT records, the processing code is triggered when the record is modified, and the record's `.VAL` field is read and reported back to the Python layer.

I hope this helps a bit.

**Reply by aswyatt** [reply]
I have a similar issue but I am still a little confused - maybe if I ask the question a different way: How do I set up an input record to be regularly scanned (i.e. it's value updated).

To be specific, I have a device (stepper motor) in which I want it's position to be read say every second. In a standard IOC I would set the SCAN field equal to "1 second" and the INP field to say a stream proto function. How is this achieved using the softIOC protocol (assume I have a python function called GetPosition() that I can call to retrieve the motors position). One method is to create a separate process within python (e.g. using asyncio or cothread) that will call the function and then set the PV value at the desired scan rate. However, is there a way to use the EPICS SCAN field or not?

**Reply by Araneidae** [reply]
Using an internal Python process is the intended way to do this.  You _could_ use a `SCAN` setting on an Out record (probably an `Action()` instance) to trigger your processing if you prefer, but I don't think the In records will work properly if you set their `SCAN` fields to anything other than `I/O Intr`.

**Reply by aswyatt** [reply]
Thanks - I wasn't sure if that was the intended methodology, or if I was missing something.

I don't want to set up an output record. since it is actually an EPICS input (doing so would lead to later confusion).
